### PR TITLE
quadrature class

### LINKED ISSUE #13: Task 11
Integration Engine

### PR DESCRIPTION
**Pull request by wdwatson2**
## Description

Implements Guass-Legendre and Gauss-Lobatto quadrature.

An instance computes the nodes and weights on $[-1, 1]$. The `integrate()` method takes a `Callable` function and maps the nodes to the interval $[a,b]$ and returns the approximation.

## Linked Issue
Fixes #13 

## Implementation Checklist
- [] My class inherits from `PolynomialBasis` (if applicable) (N/A)
- [] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.) (N/A)
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### CODE REVIEW

File: tests/test_integrator.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Test file (unchanged except for line length in one test)
+import numpy as np
+import pytest
+from numanalysislib.basis.power import PowerBasis
+from src.numanalysislib.calculus.integrator import Quadrature
+
+
+class TestQuadrature:
+
```
Comment (Lemux133) [inline tests/test_integrator.py:9]: perhaps consider add one more test for some constant function (as an edge case)?

File: tests/test_integrator.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    def test_lobatto_error(self):
+        with pytest.raises(ValueError):
+            quad = Quadrature(rule='gauss-lobatto', n_points=1)
+
+    def test_points_error(self):
+        with pytest.raises(ValueError):
+            quad = Quadrature(rule='gauss-legendre', n_points=-1)
+
+    def test_power_basis(self):
```
Comment (Lemux133) [inline tests/test_integrator.py]: a more generic test building on this would be testing that the quadrature is exact for polynomials up to degree 2n-1

**Comment by Lemux133** [comment]
## Peer Review Checklist

### 1. Mathematical Integrity
- [x] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [x] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly? #having several edge cases like infinite bounds, singular integral, etc.
- [x] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)? #created customized affine mapper for the quadrature evaluation

### 2. API & Style
- [N/A] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [x] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [x] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [x] **Coverage:** Unit tests cover typical usage and edge cases.
- [x] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [x] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

File: src/numanalysislib/calculus/integrator.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        # Compute weights using formula: 2/(n(n+1)) * 1/[L_n(x_j)]^2
+        n = degree
+        L_n = eval_legendre(n, points)
+        weights = 2.0 / (n * (n + 1)) * (1.0 / (L_n ** 2))
+
+        return points, weights
+
+    def _validate_bounds(self, a: float, b: float) -> None:
+        """
+        Raises errors if its not a valid bounded interval [a,b].
```
Comment (mtezzele) [inline src/numanalysislib/calculus/integrator.py:118]: add parameters doc

File: src/numanalysislib/calculus/integrator.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        """
+        if not np.isfinite(a) or not np.isfinite(b):
+            raise ValueError(f"Integration bounds must be finite. Got a={a}, "
+                             f"b={b}")
+        if a > b:
+            raise ValueError(f"Lower bound a={a} must be <= upper bound b={b}")
+
+    def _affine_map(self, a: float, b: float) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (mtezzele) [inline src/numanalysislib/calculus/integrator.py:133]: missing doc

File: tests/test_integrator.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        np.testing.assert_allclose(result, exact, rtol=1e-6)
+
+    def test_improper_integral_divergent(self):
+        quad = Quadrature(rule='gauss-legendre', n_points=1000)
+        result = quad.integrate(lambda x: 1/x, 0, 1)
+        quad_few = Quadrature(rule='gauss-legendre', n_points=100)
+        result_few = quad_few.integrate(lambda x: 1/x, 0, 1)
+
+        # With more points, the approximation should get larger
+        assert result > result_few
```
Comment (mtezzele) [inline tests/test_integrator.py]: Check whether you can use `np.testing.assert_array_less` or similar

**Review by mtezzele — COMMENTED** [review]
Please add a test to check that nodes and weights are as expected

File: tests/test_integrator.py (hunk 1 above)
Comment (wdwatson2) [reply tests/test_integrator.py:9]: added in most recent commit

File: tests/test_integrator.py (hunk 2 above)
Comment (wdwatson2) [reply tests/test_integrator.py]: added in most recent commit

File: src/numanalysislib/calculus/integrator.py (hunk 3 above)
Comment (wdwatson2) [reply src/numanalysislib/calculus/integrator.py:118]: added in most recent commit

File: src/numanalysislib/calculus/integrator.py (hunk 4 above)
Comment (wdwatson2) [reply src/numanalysislib/calculus/integrator.py:133]: added in most recent commit

File: tests/test_integrator.py (hunk 5 above)
Comment (wdwatson2) [reply tests/test_integrator.py]: added in most recent commit

**Comment by mtezzele** [comment]
@wdwatson2 you merged the main branch into this branch instead of doing a fetch + rebase. Please close it and open a new clean PR.
